### PR TITLE
[WIP] Add Beta CDF

### LINKED ISSUE #21553: Feature Request
If you have a question or would like help and support, please ask at our
[forums](https://discuss.pytorch.org/).

If you are submitting a feature request, please preface the title with [feature request].
If you are submitting a bug report, please fill in the following details.

## Issue description

Provide a short description.

## Code example

Please try to provide a minimal example to repro the bug.
Error messages and stack traces are also helpful.

## System Info
Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

- PyTorch or Caffe2:
- How you installed PyTorch (conda, pip, source):
- Build command you used (if compiling from source):
- OS:
- PyTorch version:
- Python version:
- CUDA/cuDNN version:
- GPU models and configuration:
- GCC version (if compiling from source):
- CMake version:
- Versions of any other relevant libraries:

### PR DESCRIPTION
**Pull request by JamesTrick**
**Summary:** Work in progress work on adding the CDF for the Beta distribution. Tests are currently failing, due to the incorrect handling in the if statements. Closes #21553

**Notes:** Beta CDF is the Incomplete Beta Function - I've taken the approach of using a continued fraction within this. I believe this is slightly different to Numpy, which can yield very small differences between Numpy (and consequently TFP).

**Help Wanted:** Currently the CDF tests are failing for the beta distribution, I'm getting a `RuntimeError: bool value of Tensor with more than one value is ambiguous` caused by the comparison when `value` is more than one value. How can I successfully handle this?

### CODE REVIEW

File: torch/distributions/utils.py [hunk 1]
```diff
@@ -85,6 +86,45 @@ def probs_to_logits(probs, is_binary=False):
     return torch.log(ps_clamped)
 
 
+def continued_fraction(a, b, value, niters=2000, tol=3.0e-11):
+    r"""
+    Evaluates the continued fraction form of the Incomplete Beta Function,
+    required for Beta CDF.
+
+    Adapted from Numerical Recipes in C, Section 6.4
+    (http://www.aip.de/groups/soe/local/numres/bookcpdf/c6-4.pdf)
```
Comment (alicanb) [inline torch/distributions/utils.py]: link doesn't work

File: torch/distributions/beta.py [hunk 2]
```diff
@@ -66,6 +66,20 @@ def log_prob(self, value):
     def entropy(self):
         return self._dirichlet.entropy()
 
+    def cdf(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+
+        else:
```
Comment (alicanb) [inline torch/distributions/beta.py]: remove the `else`

File: torch/distributions/beta.py [hunk 3]
```diff
@@ -66,6 +66,20 @@ def log_prob(self, value):
...
+            lbeta = torch.lgamma(self.concentration1 + self.concentration0) - torch.lgamma(self.concentration1) - \
+                torch.lgamma(self.concentration0) + self.concentration1 * torch.log(value) + self.concentration0 * \
+                torch.log(1.0 - value)
+            if torch.where(value < (self.concentration1 + 1.0) / (self.concentration1 + self.concentration0 + 2.0)):
+                return torch.exp(lbeta) * continued_fraction(self.concentration1, self.concentration0, value) / self.concentration1
+            else:
+                return 1.0 - torch.exp(lbeta) * continued_fraction(self.concentration0, self.concentration1, 1.0 - value) / \
+                    self.concentration0
```
Comment (alicanb) [inline torch/distributions/beta.py]: I think you want to use a pattern like this:
```
output = calculate_output_condition_is_false(input)
output[condition(input)] = calculate_output_condition_is_true(input[condition(input)])
```

File: torch/distributions/utils.py [hunk 4]
```diff
@@ -85,6 +86,45 @@ def probs_to_logits(probs, is_binary=False):
...
+        bp = bz + d * bm
+        d = -(a + em) * ((a + b) + em) * value / ((qap + tem) * (a + tem))
+        app = ap + d * az
+        bpp = bp + d * bz
+        a_old = az
+        am = ap / bpp
+        bm = bp / bpp
+        az = app / bpp
+        bz = 1.0
+        if torch.where(abs(az - a_old) < (tol * abs(az))):
```
Comment (alicanb) [inline torch/distributions/utils.py]: you might want to check the docs for torch.where https://pytorch.org/docs/stable/torch.html#torch.where

**Comment by JohnLaMaster** [comment]
How close is this to being finished? It's been 3 years and no one else has implemented anything either. I have a use case and would appreciate getting this finalized and implemented. Please let me know if you need help finishing it.

**Pull request closed without merging**